doc/plugins.rst

```diff
@@ -5,74 +5,90 @@ tox plugins
-Python packaging index:
+You can create a new tox plugin with all the bells and whistles via a `Cookiecutter`_ template (see `cookiecutter-tox-plugin <https://github.com/tox-dev/cookiecutter-tox-plugin>`_) - this will create a complete pypi-releasable, documented project with license, documentation and CI.
 
-.. code-block:: python
-
-    # content of setup.py
-    from setuptools import setup
+.. code-block:: console
 
-    if __name__ == "__main__":
-        setup(
-            name='tox-MYPLUGIN',
-            description='tox plugin decsription',
-            license="MIT license",
-            version='0.1',
-            py_modules=['tox_MYPLUGIN'],
-            entry_points={'tox': ['MYPLUGIN = tox_MYPLUGIN']},
-            install_requires=['tox>=2.0'],
-        )
+    $ pip install -U cookiecutter
+    $ cookiecutter gh:tox-dev/cookiecutter-tox-plugin
 
-If installed, the ``entry_points`` part will make tox see and integrate
-your plugin during startup.
 
-You can install the plugin for development ("in-place") via:
+Tutorial: a minimal tox plugin
+------------------------------
 
-.. code-block:: shell
+To create a working plugin you need at least a python project with a tox entry point and a python
+module implementing one or more of the pluggy based hooks tox specifies (using the
+``@tox.hookimpl`` decorator as marker).
 
-    pip install -e .
+Let us consider you want to extent tox behaviour by displaying fireworks at the end of a
+successful tox run (we won't go into the details of how to display fireworks though).
 
-and later publish it via something like:
+minimal structure:
 
-.. code-block:: shell
+.. code-block:: console
 
-    python setup.py sdist register upload
+    $ mkdir tox-fireworks
+    $ cd tox-fireworks
+    $ touch tox_fireworks.py
+    $ touch setup.py
 
-
-Writing hook implementations
-----------------------------
-
-A plugin module defines one or more hook implementation functions
-by decorating them with tox's ``hookimpl`` marker:
+contents of ``tox_fireworks.py``:
 
 .. code-block:: python
 
-    from tox import hookimpl
+    import pluggy
+
+    hookimpl = pluggy.HookimplMarker("tox")
 
     @hookimpl
     def tox_addoption(parser):
-        # add your own command line options
-
+        """Add command line option to display fireworks on request."""
 
     @hookimpl
     def tox_configure(config):
-        # post process tox configuration after cmdline/ini file have
-        # been parsed
+        """Post process config after parsing."""
+
+    @hookimpl
+    def tox_runenvreport(config):
+        """Display fireworks if all was fine and requested."""
 
-If you put this into a module and make it pypi-installable with the ``tox``
-entry point you'll get your code executed as part of a tox run.
 
+contents of ``setup.py``:
 
+.. code-block:: python
+
+    from setuptools import setup
+
+    setup(name='tox-fireworks', py_modules=['tox_fireworks'],
```

Comment:
would be more comfortable if this would follow the structure of what we generate with the cookiecutter, but at least add https://github.com/tox-dev/cookiecutter-tox-plugin/issues/4